### PR TITLE
[FEATURE] ID/PK Rendering in DataDocs

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
## What has changed?
* `_diagnostic_unexpected_table_renderer()` has been augmented to include `unexpected_index_list` in cases where ID/PK has been defined
* For SQL (for now), query to retrieve all unexpected values is included as a dropdown menu.
* Helper methods have been added to `great_expectations/render/util.py`
* Tests have been added at the `util.py`-level
* Tests have also been added at the `ValidationResultsTableContentBlockRenderer` level. 

* Follow up to #5876 and #6448. Please refer to the PRs for additional context. 

## How does it look?
The Unexpected Counts table used to look like this:
Screen Shot 2022-12-23 at 10 29 01 AM

### In this example ID/PK columns are defined as `pk_1` and `pk_2`. For SQL output the table can include a dropdown menu to show `QUERY` to retrieve full results (Pandas implementation coming soon)
Screen Shot 2022-12-23 at 10 28 16 AM

### For `Pandas` the unexpected indices output can include default output, without having to define ID/PK columns. The column is called `Index` by default. 
Screen Shot 2022-12-23 at 10 29 28 AM

### When there are a LOT of unexpected indices (ie more than 10), we truncate and output `...` 
Screen Shot 2022-12-23 at 11 13 18 AM

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.

### CODE REVIEW

File: great_expectations/render/renderer/content_block/validation_results_table_content_block.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ def row_generator_fn(
             if unexpected_statement:
                 expectation_string_cell += unexpected_statement
             if unexpected_table:
-                expectation_string_cell.append(unexpected_table)
+                expectation_string_cell += unexpected_table
```
Comment (Shinnnyshinshin) [inline great_expectations/render/renderer/content_block/validation_results_table_content_block.py:200]: Needed for appending `query` dropdown

File: tests/render/test_render_ValidationResultsTableContentBlockRenderer.py [hunk 2]
```diff
@@ -1098,3 +1186,201 @@ def test_ValidationResultsTableContentBlockRenderer_get_status_cell(
...
+        renderer_type=LegacyDiagnosticRendererType.UNEXPECTED_TABLE,
+    )[1](result=evr_id_pk_pandas)
+    assert rendered_value[0].to_json_dict() == {
+        "content_block_type": "table",
+        "header_row": ["Sampled Unexpected Values", "Count", "Index"],
+        "styling": {"body": {"classes": ["table-bordered", "table-sm", "mt-3"]}},
+        "table": [
+            ["giraffe", 1, "3"],
+            ["lion", 1, "4"],
+            ["zebra", 1, "5 6 7 8 9 10 11 12 13 14 ..."],
```
Comment (anthonyburdi) [inline tests/render/test_render_ValidationResultsTableContentBlockRenderer.py]: :heart:

Comment (anthonyburdi) [reply tests/render/test_render_ValidationResultsTableContentBlockRenderer.py]: Love the test of truncating with ellipsis.

File: tests/render/test_util.py [hunk 3]
```diff
@@ -284,3 +288,284 @@ def find_code_in_notebook(
...
+        unexpected_index_column_names=unexpected_index_column_names,
+        partial_unexpected_counts=partial_unexpected_counts,
+    )
+    assert list(res) == unexpected_index_column_names
+    assert res.iloc[0].tolist() == ["3"]
+    assert res.iloc[1].tolist() == ["4"]
+    assert res.iloc[2].tolist() == ["5 6 7 8"]
+
+
+def test_convert_unexpected_indices_to_df_multiple_with_truncation():
```
Comment (anthonyburdi) [inline tests/render/test_util.py:346]: :heart:

File: great_expectations/render/util.py [hunk 4]
```diff
@@ -209,3 +212,219 @@ def handle_strict_min_max(params: dict) -> (str, str):
...
+        header_row = ["Unexpected Value", "Count"]
+    else:
+        header_row = ["Sampled Unexpected Values", "Count"]
+    return header_row, table_rows
+
+
+def build_count_and_index_table(
+    partial_unexpected_counts: List[dict],
+    unexpected_index_list: List[dict],
+    unexpected_count: Optional[int] = None,
```
Comment (anthonyburdi) [inline great_expectations/render/util.py]: It looks like `unexpected_count` is not optional (compared to an int in the function)

File: great_expectations/render/util.py [hunk 5]
```diff
@@ -209,3 +212,219 @@ def handle_strict_min_max(params: dict) -> (str, str):
...
+    # https://www.geeksforgeeks.org/apply-function-to-every-row-in-a-pandas-dataframe/
+    for column in unexpected_index_column_names:
+        filtered_unexpected_indices[column] = filtered_unexpected_indices[column].apply(
+            lambda row: filter_list_of_indices(row)
+        )
+
+    return filtered_unexpected_indices
+
+
+def filter_list_of_indices(indices: List[Union[int, str]], max_index: int = 10) -> str:
```
Comment (anthonyburdi) [inline great_expectations/render/util.py]: What about changing the name to `truncate_list_of_indices`?

File: great_expectations/render/util.py [hunk 6]
```diff
@@ -209,3 +212,219 @@ def handle_strict_min_max(params: dict) -> (str, str):
...
+        max_index: Maximum number of indices to display before showing "..." in the column
+
+    Returns:
+        string of indices that are joined using ` `
+
+    """
+    if len(indices) > max_index:
+        indices = indices[:max_index]
+        indices.append("...")
+    return " ".join(map(str, indices))
```
Comment (anthonyburdi) [inline great_expectations/render/util.py]: Should we join with a comma `,` instead of a space to match Trang's design?

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Looks really good! Just a few small requested changes.

File: great_expectations/render/util.py (hunk 4 above)
Comment (Shinnnyshinshin) [reply great_expectations/render/util.py]: Thanks for catching this :)

File: great_expectations/render/util.py (hunk 5 above)
Comment (Shinnnyshinshin) [reply great_expectations/render/util.py]: 👍🏼

File: great_expectations/render/util.py (hunk 4 above)
Comment (Shinnnyshinshin) [reply great_expectations/render/util.py]: made the change in `build_count_table` as well

**Review by anthonyburdi — APPROVED** [review]
Thank you for making those changes!

**Comment by Shinnnyshinshin** [comment]
> Thank you for making those changes!

Thank you very very much. I'm going to add a little bit more detail to the PR description + some screenshots of the rendering, and set automerge 🙇🏼‍♂️

**Pull request merged**